go/types/simplify.go

```diff
@@ -0,0 +1,225 @@
+// e. all union types reachable from it also fulfill b-e
+//
+// The simplification is created roughly as follows:
+//
+// - The input types are deduplicated
+// - Any unions in the input set are "flattened" into the input set
+// - The inputs are grouped into categories:
+//    - ref
+//    - list
+//    - set
+//    - map
+//    - struct, by name (each unique struct name will have its own group)
+// - The ref, set, and list groups are collapsed like so:
+//     {Ref<A>,Ref<B>,...} -> Ref<A|B|...>
+// - The map group is collapsed like so:
+//     {Map<K1,V1>|Map<K2,V2>...} -> Map<K1|K2,V1|V2>
+// - Each struct group is collapsed like so:
+//     {struct{foo:number,bar:string}, struct{bar:bool, baz:blob}} ->
+//       struct{bar:string|blob}
+//
+// Anytime any of the above cases generates a union as output, the same process
+// is applied to that union recursively.
+func makeSimplifiedUnion(in ...*Type) *Type {
+	d.Chk.True(len(in) > 0)
+
+	ts := typeset{}
+	for _, t := range in {
+		// De-cycle so that we handle cycles explicitly below. Otherwise, we would implicitly crawl
+		// cycles and recurse forever.
+		t := ToUnresolvedType(t)
+		ts[t] = struct{}{}
+	}
+
+	r := makeSimplifiedUnionImpl(ts)
+
+	if r.HasUnresolvedCycle() {
+		r = resolveStructCycles(r, nil)
+	}
+
+	return r
+}
+
+// typeset is a helper that aggregates the unique set of input types for this algorithm, flattening
+// any unions recursively.
+type typeset map[*Type]struct{}
+
+func (ts typeset) Add(t *Type) {
+	switch t.Kind() {
+	case UnionKind:
+		for _, et := range t.Desc.(CompoundDesc).ElemTypes {
+			ts.Add(et)
+		}
+	default:
+		ts[t] = struct{}{}
+	}
+}
+
+func newTypeset(t ...*Type) typeset {
+	ts := typeset{}
+	for _, t := range t {
+		ts.Add(t)
+	}
+	return ts
+}
+
+// makeSimplifiedUnionImpl is an implementation detail of MakeSimplifiedUnion.
+// Warning: Do not call this directly. It assumes its input has been
+// de-cycled using ToUnresolvedType() and will inifinitely recurse otherwise
+// on cyclic types otherwise.
+func makeSimplifiedUnionImpl(in typeset) *Type {
+	type how struct {
+		k NomsKind
+		n string
+	}
+
+	out := []*Type{}
+	groups := map[how]typeset{}
+	for t, _ := range in {
+		var h how
+		switch t.Kind() {
+		case RefKind, SetKind, ListKind, MapKind:
+			h = how{k: t.Kind()}
+		case StructKind:
+			h = how{k: t.Kind(), n: t.Desc.(StructDesc).Name}
+		default:
+			out = append(out, t)
+			continue
+		}
+		g := groups[h]
+		if g == nil {
+			g = typeset{}
+			groups[h] = g
+		}
+		g.Add(t)
+	}
+
+	for h, ts := range groups {
+		if len(ts) == 1 {
+			for t, _ := range ts {
+				out = append(out, t)
+			}
+			continue
+		}
+
+		var r *Type
+		switch h.k {
+		case RefKind:
+			r = simplifyRefs(ts)
+		case SetKind:
+			r = simplifySets(ts)
+		case ListKind:
+			r = simplifyLists(ts)
+		case MapKind:
+			r = simplifyMaps(ts)
+		case StructKind:
+			r = simplifyStructs(h.n, ts)
+		}
+		out = append(out, r)
+	}
+
+	if len(out) == 1 {
```

Comment:
Yeah, it's kind of a bummer to call `MakeUnionType` at all, because it does a ton of work again. But I tried to lift the code I needed out of there and bugs happened.